src/server/json_family.cc

```diff
@@ -89,82 +134,163 @@ OpResult<string> OpGet(const OpArgs& op_args, string_view key, vector<pair<strin
-
-  error_code ec;
-  json_decoder<json> decoder;
-  const PrimeValue& pv = it_res.value()->second;
+OpResult<vector<OptSizeT>> OpObjLen(const OpArgs& op_args, string_view key,
+                                    JsonExpression expression) {
+  OpResult<json> result = GetJson(op_args, key);
+  if (!result) {
+    return result.status();
+  }
 
-  string val = GetString(op_args.shard, pv);
-  basic_json_parser<char> parser(basic_json_decode_options<char>{}, &JsonErrorHandler);
+  vector<OptSizeT> vec;
+  auto cb = [&vec](const string_view& path, const json& val) {
+    if (val.is_object()) {
+      vec.emplace_back(val.object_value().size());
+    } else {
+      vec.emplace_back(nullopt);
+    }
+  };
 
-  parser.update(val);
-  parser.finish_parse(decoder, ec);
+  expression.evaluate(*result, cb);
+  return vec;
+}
 
-  if (!decoder.is_valid()) {
-    return OpStatus::SYNTAX_ERR;
+OpResult<vector<OptSizeT>> OpArrLen(const OpArgs& op_args, string_view key,
+                                    JsonExpression expression) {
+  OpResult<json> result = GetJson(op_args, key);
+  if (!result) {
+    return result.status();
   }
 
   vector<OptSizeT> vec;
   auto cb = [&vec](const string_view& path, const json& val) {
-    if (val.is_string()) {
-      vec.emplace_back(val.as_string_view().size());
+    if (val.is_array()) {
+      vec.emplace_back(val.array_value().size());
+    } else {
+      vec.emplace_back(nullopt);
+    }
+  };
+
+  expression.evaluate(*result, cb);
+  return vec;
+}
+
+OpResult<vector<OptBool>> OpToggle(const OpArgs& op_args, string_view key, string_view path) {
+  OpResult<json> result = GetJson(op_args, key);
+  if (!result) {
+    return result.status();
+  }
+
+  vector<OptBool> vec;
+  auto cb = [&vec](const string& path, json& val) {
+    if (val.is_bool()) {
+      bool current_val = val.as_bool() ^ true;
+      val = current_val;
+      vec.emplace_back(current_val);
     } else {
       vec.emplace_back(nullopt);
     }
   };
 
-  expression.evaluate(decoder.get_result(), cb);
+  json j = result.value();
+  error_code ec = JsonReplace(j, path, cb);
+  if (ec) {
+    VLOG(1) << "Failed to evaulate expression on json with error: " << ec.message();
+    return OpStatus::SYNTAX_ERR;
+  }
+
+  SetString(op_args, key, j.as_string());
   return vec;
 }
 
-} // namespace
+}  // namespace
+
+void JsonFamily::Toggle(CmdArgList args, ConnectionContext* cntx) {
+  string_view key = ArgS(args, 1);
+  string_view path = ArgS(args, 2);
+
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    return OpToggle(t->GetOpArgs(shard), key, path);
+  };
+
+  DVLOG(1) << "Before Get::ScheduleSingleHopT " << key;
+  Transaction* trans = cntx->transaction;
+  OpResult<vector<OptBool>> result = trans->ScheduleSingleHopT(move(cb));
+
+  if (result) {
+    DVLOG(1) << "JSON.TOGGLE " << trans->DebugId() << ": " << key;
+    if (result->empty()) {
+      (*cntx)->SendNullArray();
+    } else {
+      (*cntx)->StartArray(result->size());
```

Comment:
there are three repetitions of the same code of sending a long array.
We can factor them out in the next changes.

Reply:
I put it as a requirement for the next PR.